File: serve_build.mjs
import { createServer } from "node:http";
import { readFile } from "node:fs/promises";
import { extname, join, normalize } from "node:path";

const root = normalize(join(process.cwd(), "build", "web"));
const port = 53217;
const types = {
  ".html": "text/html; charset=utf-8",
  ".js": "text/javascript; charset=utf-8",
  ".css": "text/css; charset=utf-8",
  ".json": "application/json; charset=utf-8",
  ".png": "image/png",
  ".ico": "image/x-icon",
  ".svg": "image/svg+xml",
  ".wasm": "application/wasm",
  ".otf": "font/otf",
  ".ttf": "font/ttf",
};

createServer(async (request, response) => {
  const url = new URL(request.url, `http://localhost:${port}`);
  const path = url.pathname === "/" ? "/index.html" : url.pathname;
  const target = normalize(join(root, decodeURIComponent(path)));

  if (!target.startsWith(root)) {
    response.writeHead(403);
    response.end("Forbidden");
    return;
  }

  try {
    const data = await readFile(target);
    response.writeHead(200, {
      "content-type": types[extname(target)] || "application/octet-stream",
      "cache-control": "no-store",
    });
    response.end(data);
  } catch {
    const fallback = await readFile(join(root, "index.html"));
    response.writeHead(200, {
      "content-type": "text/html; charset=utf-8",
      "cache-control": "no-store",
    });
    response.end(fallback);
  }
}).listen(port, () => {
  console.log(`Pixel Painter build is running at http://localhost:${port}`);
});

File: serve_build.dart
import 'dart:io';

Future<void> main() async {
  final root = Directory('build/web').absolute;
  final server = await HttpServer.bind(InternetAddress.loopbackIPv4, 53217);
  stdout.writeln(
    'Pixel Painter build is running at http://localhost:${server.port}',
  );

  await for (final request in server) {
    final uriPath = request.uri.path == '/' ? '/index.html' : request.uri.path;
    final requestedPath = uriPath
        .substring(1)
        .split('/')
        .where((segment) => segment.isNotEmpty)
        .join(Platform.pathSeparator);
    final target = File('${root.path}${Platform.pathSeparator}$requestedPath');

    if (requestedPath.contains('..') ||
        !target.absolute.path.startsWith(root.path)) {
      request.response.statusCode = HttpStatus.forbidden;
      await request.response.close();
      continue;
    }

    final file = await target.exists()
        ? target
        : File.fromUri(root.uri.resolve('index.html'));
    request.response.headers
      ..contentType = _contentType(file.path)
      ..set(HttpHeaders.cacheControlHeader, 'no-store');
    await request.response.addStream(file.openRead());
    await request.response.close();
  }
}

ContentType _contentType(String path) {
  if (path.endsWith('.html')) {
    return ContentType.html;
  }
  if (path.endsWith('.js')) {
    return ContentType('text', 'javascript', charset: 'utf-8');
  }
  if (path.endsWith('.css')) {
    return ContentType('text', 'css', charset: 'utf-8');
  }
  if (path.endsWith('.json')) {
    return ContentType.json;
  }
  if (path.endsWith('.png')) {
    return ContentType('image', 'png');
  }
  if (path.endsWith('.svg')) {
    return ContentType('image', 'svg+xml');
  }
  if (path.endsWith('.wasm')) {
    return ContentType('application', 'wasm');
  }
  return ContentType.binary;
}
